### PR TITLE
voice assistant pkg

### PR DESCRIPTION
**Pull request by barzanisar**

### CODE REVIEW

File: trailbot_interfaces/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (jacobdeery) [inline trailbot_interfaces/CMakeLists.txt:1]: Did this package make it in here by mistake? I think this should just be the voice assistant package.

File: voice_assistant_pkg/LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (jacobdeery) [inline voice_assistant_pkg/LICENSE:1]: We can probably remove the individual package LICENSE files and just keep a single one at the top level of the repository.

File: voice_assistant_pkg/README.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+## 1.2. Install python packages
+
+```bash
+# [Optional] Activate virtual env
+source ~/ros2_ws/src/colcon_venv/venv/bin/activate
+
+# Install required packages
+pip install openai
+pip install SpeechRecognition
```
Comment (jacobdeery) [inline voice_assistant_pkg/README.md:79]: Can you create a `requirements.txt` file with all pip dependencies?

File: voice_assistant_pkg/README.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+sudo apt install espeak
+sudo apt-get update -y
+sudo apt-get install -y portaudio19-dev
+pip install PyAudio
+
+# for elevenlabs
+pip install elevenlabs
+pip install elevenlabslib
+pip install pydub
+sudo apt-get install ffmpeg
```
Comment (jacobdeery) [inline voice_assistant_pkg/README.md:92]: Please create a `requirements.txt` file with all pip dependencies

File: voice_assistant_pkg/package.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>voice_assistant_pkg</name>
```
Comment (jacobdeery) [inline voice_assistant_pkg/package.xml:4]: conventionally, packages don't have `pkg` in the name (so just `voice_assistant`)

File: voice_assistant_pkg/voice_assistant_pkg/conversation_handler.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import os
+import json
+
+class ConversationHandler:
+    def __init__(self, messages, save_foldername):
+        """AI is creating summary for __init__
```
Comment (jacobdeery) [inline voice_assistant_pkg/voice_assistant_pkg/conversation_handler.py:6]: 👀

File: voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+        Args:
+            audio : Speech recognizer object that contains user's input 
+
+        Returns:
+            user_input (str): transcribed voice command
+        """
+        with open('speech.wav','wb') as f:
+            f.write(audio.get_wav_data())
+        speech = open('speech.wav', 'rb')
```
Comment (jacobdeery) [inline voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py:28]: Will this be saved inside the workspace?

File: voice_assistant_pkg/voice_assistant_pkg/voice_assistant.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        self.messages.append({"role": "assistant", "content": response})
+        print(f"\n{response}\n")
+        self.conversation_handler.save_inprogress(self.messages)
+        
+        self.speak(f'{response}')
+            
+    
+    def process_user_input(self, user_input):
+        # Keep chatting until user goes silent
+        while user_input is not None:
```
Comment (jacobdeery) [inline voice_assistant_pkg/voice_assistant_pkg/voice_assistant.py:240]: Should we have an `rclpy.ok()` in here too?

**Review by jacobdeery — COMMENTED** [review]
Overall this looks great! Just a few minor comments.

Could you also hit the code with a formatter like `autopep8` to clean up some of the whitespace?

File: trailbot_interfaces/CMakeLists.txt (hunk 1 above)
Comment (barzanisar) [reply trailbot_interfaces/CMakeLists.txt:1]: voice_assistant_pkg depends on trailbot interfaces so voice_assistant_pkg branch branches out of trailbot interfaces branch. If you just pull voice_assistant_pkg branch it won't build.

File: voice_assistant_pkg/config/params.yaml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/voice_assistant_node:
+  ros__parameters:
+    snack_options: ['chips', 'chocolate', 'candies', 'nuts'] #TODO: handle chocolates or candy
+    available_snacks_quantity: [5, 5, 5, 5]
+    exit_cmd_options: ['bye', 'bubye', 'adios', 'ciao', 'no thanks', 'no thank you', 'see you', 'take care']
+    speech_recognizer:
+      energy_threshold: 200
+      mic_device_index: 12
+      phrase_time_limit: 4 # maximum number of seconds speech recognizer will allow a phrase to continue before stopping and returning the part of the phrase processed before the time limit was reached. The resulting audio will be the phrase cut off at the time limit. If ``phrase_timeout`` is ``None``, there will be no phrase time limit.
+      timeout: 4 # maximum number of seconds speech recognizer will wait for a phrase to start before giving up. If ``timeout`` is ``None``, there will be no wait timeout.
```
[the remark this answers is not in the document]
Comment (barzanisar) [reply voice_assistant_pkg/config/params.yaml]: done!

File: voice_assistant_pkg/LICENSE (hunk 2 above)
Comment (barzanisar) [reply voice_assistant_pkg/LICENSE:1]: done in barza/trailbot_interfaces PR

File: voice_assistant_pkg/voice_assistant_pkg/conversation_handler.py (hunk 6 above)
Comment (barzanisar) [reply voice_assistant_pkg/voice_assistant_pkg/conversation_handler.py:6]: oops, edited this description

File: voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py (hunk 7 above)
Comment (barzanisar) [reply voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py:28]: yes, where else should we save it? Should I create a user-defined path parameter to save it?

File: voice_assistant_pkg/README.md (hunk 4 above)
Comment (barzanisar) [reply voice_assistant_pkg/README.md]: done

File: voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py (hunk 7 above)
Comment (jacobdeery) [reply voice_assistant_pkg/voice_assistant_pkg/speech_recognizer.py]: This is fine for now, but probably better to save it to a temp directory unless we want to keep it around long-term

**Review by jacobdeery — APPROVED** [review]
Looks good, go ahead and merge after resolving conflicts

**Pull request merged**